### ISSUE TITLE
Automated Tests

### ISSUE THREAD

**Issue by mdhaber** [MrShoenel/metrics-as-scores#4]
One of the items in the JOSS review criteria #1 is:

> **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?

I see that there are pytest tests, but there are no instructions for executing them. 
- Can they be executed if the package is installed using pip? For instance, ```import scipy; scipy.test()``` executes SciPy tests.
- How should the tests be run if the package is installed for development? For instance, running SciPy's tests is not as simple as executing `pytest` at the command line; there are developer tools `runtests.py` and now`dev.py` that should be used to run the test suite. 

Please add documentation so that I can run the test suite. After #2 is complete, I will be able to assess whether the test suite is adequate. Thanks!

**Comment by MrShoenel** [comment]
Automated tests were added. All crucial parts are covered, having achieved coverage of ~90%.

- It is not intended for Metrics As Scores to run tests if installed as PyPI package.
- There are instructions for how to run the test suite once the package is installed for development (see README.md)

The commands are (in this example, having activated the virtual environment in Windows):

```shell
# Runs the tests and prints coverage
(venv) C:\metrics-as-scores>poe test
# Writes reports to the local directory htmlcov
(venv) C:\metrics-as-scores>poe cov
```

**Comment by mdhaber** [comment]
Using conda, I created a virtual environment and installed Python 3.10 and Poetry (from conda-forge). `poetry install` failed, so I tried uninstalling the conda version of Poetry and using `pip install poetry` as instructed. `poetry install` appears to succeed, and `poe test` yields:

<details>

```
(mas) C:\Users\matth\OneDrive\Desktop\metrics-as-scores>poe test
Poe => pytest --no-cov
================================================= test session starts =================================================
platform win32 -- Python 3.10.9, pytest-7.2.1, pluggy-1.0.0
rootdir: C:\Users\matth\OneDrive\Desktop\metrics-as-scores, configfile: pyproject.toml, testpaths: ./test
plugins: anyio-3.6.2, cov-4.0.0
collected 30 items

test\cli\test_helpers.py .....                                                                                   [ 16%]
test\data\test_Dataset.py ....                                                                                   [ 30%]
test\data\test_pregenerate.py ...                                                                                [ 40%]
test\data\test_pregenerate_distns.py .                                                                           [ 43%]
test\data\test_pregenerate_fit.py ..                                                                             [ 50%]
test\distribution\test_distribution.py .......                                                                   [ 73%]
test\distribution\test_fitting.py ...                                                                            [ 83%]
test\distribution\test_fitting_problems.py ..                                                                    [ 90%]
test\tools\test_funcs.py ..                                                                                      [ 96%]
test\tools\test_lazy.py .                                                                                        [100%]

================================================== warnings summary ===================================================
src\metrics_as_scores\distribution\distribution.py:36
  C:\Users\matth\OneDrive\Desktop\metrics-as-scores\src\metrics_as_scores\distribution\distribution.py:36: DeprecationWarning: invalid escape sequence '\m'
    """

src\metrics_as_scores\distribution\distribution.py:312
  C:\Users\matth\OneDrive\Desktop\metrics-as-scores\src\metrics_as_scores\distribution\distribution.py:312: DeprecationWarning: invalid escape sequence '\i'
    """

src\metrics_as_scores\tools\funcs.py:83
  C:\Users\matth\OneDrive\Desktop\metrics-as-scores\src\metrics_as_scores\tools\funcs.py:83: DeprecationWarning: invalid escape sequence '\d'
    return [int(t) if t.isdigit() else t.lower() for t in split('(\d+)', f'{s}')]

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================================== 30 passed, 3 warnings in 21.11s ===========================================
```

</details>

The output for `poe cov` is:

<details>

```
(mas) C:\Users\matth\OneDrive\Desktop\metrics-as-scores>poe cov
Poe => pytest --cov-report html:htmlcov
================================================= test session starts =================================================
platform win32 -- Python 3.10.9, pytest-7.2.1, pluggy-1.0.0
rootdir: C:\Users\matth\OneDrive\Desktop\metrics-as-scores, configfile: pyproject.toml, testpaths: ./test
plugins: anyio-3.6.2, cov-4.0.0
collected 30 items

test\cli\test_helpers.py .....                                                                                   [ 16%]
test\data\test_Dataset.py ....                                                                                   [ 30%]
test\data\test_pregenerate.py ...                                                                                [ 40%]
test\data\test_pregenerate_distns.py .                                                                           [ 43%]
test\data\test_pregenerate_fit.py ..                                                                             [ 50%]
test\distribution\test_distribution.py .......                                                                   [ 73%]
test\distribution\test_fitting.py ...                                                                            [ 83%]
test\distribution\test_fitting_problems.py ..                                                                    [ 90%]
test\tools\test_funcs.py ..                                                                                      [ 96%]
test\tools\test_lazy.py .                                                                                        [100%]

---------- coverage: platform win32, python 3.10.9-final-0 -----------
Name                                                     Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------------
src\metrics_as_scores\__init__.py                            9      0   100%
src\metrics_as_scores\__version__.py                        14      0   100%
src\metrics_as_scores\cli\__init__.py                        0      0   100%
src\metrics_as_scores\cli\helpers.py                        55      0   100%
src\metrics_as_scores\data\__init__.py                       0      0   100%
src\metrics_as_scores\data\pregenerate.py                   94      0   100%
src\metrics_as_scores\data\pregenerate_distns.py            29      0   100%
src\metrics_as_scores\data\pregenerate_fit.py               65      0   100%
src\metrics_as_scores\distribution\__init__.py               0      0   100%
src\metrics_as_scores\distribution\distribution.py         416      0   100%
src\metrics_as_scores\distribution\fitting.py              178      0   100%
src\metrics_as_scores\distribution\fitting_problems.py      88      0   100%
src\metrics_as_scores\tools\__init__.py                      0      0   100%
src\metrics_as_scores\tools\funcs.py                        37      0   100%
src\metrics_as_scores\tools\lazy.py                         91      0   100%
--------------------------------------------------------------------------------------
TOTAL                                                     1076      0   100%
Coverage HTML written to dir htmlcov
```

</details>

I see that the files that are tested get 100% coverage, and the tests I checked (e.g. test_distribution.py) seem to be looking for reasonable things. There are a few questions:
1. Some files aren't tested. Why are these files not important to test?
2. The criterion also asks whether there are "manual steps described o that the functionality of the software can be verified". I imagine that this would be applicable to the TUI and web app. 

I think 2 can be satisfied with examples (#5), so I'll close this if the answer to #1 seems appropriate.

**Comment by MrShoenel** [comment]
The warnings should have disappeared by now, as I made sure that strings using regular expressions are now raw strings.

The files that are excluded from testing are the following:
- The CLI's workflows (not the helpers though): The primary reason is that the workflows only collect user input, which is then used to instantiate and instrument the otherwise completely tested code. So even invalid input would be caught. Also, all user input is either validated within the workflows or the user is presented with a set of hard-coded choices (which will still be validated further down, e.g., when choosing random variables).
- The web app: The web app makes use of tested code, too. I have now also improved the handling of datasets, in that the web app makes consistency checks before loading it. The web app itself has rich error handling, through the console/CLI, as well as showing errors in the GUI.

I will add usage examples (*how*) as notebooks soon for #5.
